### PR TITLE
Synchronized loss before backward and added tasks-per-node as paramet…

### PR DESCRIPTION
**Pull request by gtnaila**
…er to train script.

### CODE REVIEW

File: infra/cray_infra/training/launch_training_job.py
Comment (greg1232) [inline infra/cray_infra/training/launch_training_job.py:1]: The number of tasks per node should be specified in sbatch not mpirun.  mpirun automatically picks it up from the PMI interface when SLURM launches the job

File: ml/cray_megatron/megatron/training_loop.py [hunk 1]
```diff
@@ -108,6 +109,16 @@ def training_step(self, batch):
...
+        
+        # Synchronize loss across all ranks
+        comm = MPI.COMM_WORLD
+        local_loss = loss.detach().clone()
+        global_loss = torch.tensor([local_loss.item()], device=device)
+        comm.Allreduce(MPI.IN_PLACE, global_loss, op=MPI.SUM)
+        avg_loss = global_loss.item() / comm.Get_size()
+
+        # Scale the original loss to match the average
+        loss = loss * (avg_loss / local_loss.item())
```
Comment (greg1232) [inline ml/cray_megatron/megatron/training_loop.py:121]: Not sure why line 121 is needed.  I think the final loss should just be the avg_loss

Comment (gtnaila) [reply ml/cray_megatron/megatron/training_loop.py:120]: I ran into an issue if I don't do it this way. Basically, if I update the loss directly, it loses its connection to the graph and then cannot be used in the backward() function. By doing this scaling trick, I maintain that connection to the graph.

File: infra/cray_infra/training/launch_training_job.py [hunk 2]
```diff
@@ -141,10 +141,7 @@ def get_train_job_entrypoint(train_args: Dict):
 
 def get_tasks_per_node(train_args: Dict):
     requested_gpu_count = train_args.get("gpus", 1)
-
-    max_gpu_count = get_max_gpu_count_from_slurm()
-
-    return min(requested_gpu_count, max_gpu_count)
```
Comment (greg1232) [inline infra/cray_infra/training/launch_training_job.py:147]: Why doesn't it use the max gpu count from slurm?  What if the user requests more GPUs than are available?

File: scripts/train_job_entrypoint.sh [hunk 3]
```diff
@@ -17,4 +17,4 @@ LOCAL_DIRECTORY="$( cd "$( dirname "${CRAY_TRAINING_JOB_CONFIG_PATH}" )" >/dev/n
 # Put the current ml directory in the python path so that the modules can be imported
 export PYTHONPATH=$LOCAL_DIRECTORY/ml:$PYTHONPATH
 
-mpirun --allow-run-as-root python $LOCAL_DIRECTORY/ml/cray_megatron/main.py $*
+mpirun --allow-run-as-root -np $SLURM_NTASKS_PER_NODE --oversubscribe python $LOCAL_DIRECTORY/ml/cray_megatron/main.py $*
```
[the remark this answers is not in the document]
Comment (gtnaila) [reply scripts/train_job_entrypoint.sh:20]: The SLURM_NTASKS_PER_NODE environment variable is created when a slurm job is started and allocated resources. It is set by slurm as part of the job's environment and becomes available to the job script and any processes launched within the job. I believe it is created by the sbatch command.

File: infra/cray_infra/training/launch_training_job.py (hunk 2 above)
Comment (gtnaila) [reply infra/cray_infra/training/launch_training_job.py:147]: Reverted.

File: scripts/train_job_entrypoint.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (greg1232) [reply scripts/train_job_entrypoint.sh:20]: I believe that mpirun can read these ENV variables directly without needing them to be passed as CLI variables.

[the remark this answers is not in the document]
Comment (gtnaila) [reply scripts/train_job_entrypoint.sh]: You are right. I tested it. Removed the ENV.
